File: examples/perl/gmrpl_cb.pl
#
# Copyright 2007-2016 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.
#

# 
# gmrpl_cb.pl
#
# Another perl example demonstrating the reply/request abilities of 
# GMSEC.  This example involves the use of Callback functionality,
# which allows for code to be performed anytime a Message is 
# received.
#


use strict;
use Util;

use libgmsec_perl;

*isa = \&UNIVERSAL::isa;


# Define the RequestCallback Reply Callback Package
{
	package RequestCallback;
	use base 'libgmsec_perl::Callback';

	*isa = \&UNIVERSAL::isa;

	sub onMessage
	{
		my ($self, $connection, $message) = @_;

		eval
		{
			print "In onMessage!\n";
			# Display the XML representation of the received message
			libgmsec_perl::LogInfo("[RequestCallback:onMessage] Received:\n" . $message->toXML());

			# Construct Reply subject
			my $component = 0;

			eval
			{
				$component = $message->getStringField("COMPONENT")->getValue();
			};
			if (isa($@, 'libgmsec_perl::Exception'))
			{
				libgmsec_perl::LogWarning("COMPONENT field is not available: " . $@->what());
			}
			elsif($@)
			{
				libgmsec_perl::LogError($@);
			}

			# Set Status Code.  See API Interface Specification for available status codes.
			my $status_code = "1"; # 1 = Acknowledgement

			# Create reply message
			my $reply_subject = "GMSEC.MISSION.SAT_ID.RESP." . ($component ? $component : "UNKNOWN-COMPONENT") . "." . $status_code;

			my $reply = libgmsec_perl::Message->new($reply_subject, $libgmsec_perl::Message::REPLY);

			# Add fields to the reply message
			$reply->addField("COMPONENT", "GMRPL");
			$reply->addField("ANSWER", "Sure looks like it!");

			# Display XML representation of reply message
			libgmsec_perl::LogInfo("Prepared Reply:\n" . $reply->toXML());

			# Send Reply
			$connection->reply($message, $reply);
		};
		if (isa($@, 'libgmsec_perl::Exception'))
		{
			libgmsec_perl::LogError("[RequestCallback::onMessage] " . $@->what());
		}
		elsif($@)
		{
			libgmsec_perl::LogError($@);
		}
	}
} # end RequestCallback definition

{
	my $task = bless({});
	Util::driver($task, @ARGV);
}


sub printUsage
{
	my ($self, $config) = @_;
	my $hasConnectionType = Util::hasConnectionType($config);
	return 0 if ($hasConnectionType);

	print "\nusage: perl gmrpl.pl connectiontype=<middleware> [ <parameter>=<value> ]" .
	"\n\n\tNote that the parameter 'connectiontype' is required." .
	"\n\tThe rest of other parameters are optional." .
	"\n\n\tserver=<server name> (required if middleware is not bolt/MB locally)" .
	"\n\tsubject=<subject name>" .
	"\n\tmsg_timeout_ms=<message timeout period (milliseconds)>" .
	"\n\tprog_timeout_s=<program timeout period (seconds)>" .
	"\n\tloglevel=<log level>" .
	"\n\tcfgfile=<config_filename.xml>" .
	"\n\nFor more information, see API User's Guide\n";

	return 1;
}


sub run
{
	my ($self, $config) = @_;
	$self->{CONFIG} = $config;

	my @subjects = Util::determineSubjects($config);
	my $msg_timeout_ms = Util::getInteger($config, 'MSG_TIMEOUT_MS', $libgmsec_perl::GMSEC_WAIT_FOREVER);
	my $prog_timeout_s = Util::getInteger($config, 'PROG_TIMEOUT_S', $libgmsec_perl::GMSEC_WAIT_FOREVER);

	# Output GMSEC API version
	libgmsec_perl::LogInfo(libgmsec_perl::Connection::getAPIVersion());

	eval
	{
		# Create the Connection
		my $connection = libgmsec_perl::Connection::create($config);
		$self->{CONNECTION} = $connection;

		# Connect
		$connection->connect();

		# Output middleware version
		libgmsec_perl::LogInfo("Middleware version = " . $connection->getLibraryVersion());

		my $rcb = RequestCallback->new();
		$rcb->DISOWN();

		# Output example program-specific information
		for my $subject (@subjects)
		{
			libgmsec_perl::LogInfo("Subscribing to " . $subject);
			push(@{$self->{SUBINFO}}, $connection->subscribe($subject, $rcb));
		}

		my $done = 0;
		my $count = 0;
		my $elapsedTime = 0;
		my $currentTime = 0;
		my $prevTime = 0;

		while (!$done)
		{
			if ($prog_timeout_s != $libgmsec_perl::GMSEC_WAIT_FOREVER && $elapsedTime >= $prog_timeout_s)
			{
				libgmsec_perl::LogInfo("Program Timeout Reached!");
				last;
			}

			# Receive next message
			my $message = $connection->receive($msg_timeout_ms);

			if ($prog_timeout_s != $libgmsec_perl::GMSEC_WAIT_FOREVER)
			{
				$currentTime = libgmsec_perl::TimeUtil->getCurrentTime_s();
				$elapsedTime += ($currentTime - $prevTime);
				$prevTime = $currentTime;
			}

			if (!$message)
			{
				libgmsec_perl::LogInfo("timeout occurred");
				last;
			}
			else
			{
				# Display the XML representation of the received message
				libgmsec_perl::LogInfo("Received:\n" . $message->toXML());

				# dispatch callbacks
				$connection->dispatch($message);

				if ($message->getSubject() eq "GMSEC.TERMINATE")
				{
					$done = 1;
				}

				# Destroy received message
				$connection->release($message);
			}
		}
	};
	if (isa($@, 'libgmsec_perl::Exception'))
	{
		libgmsec_perl::LogError($@->what());
	}
	elsif($@)
	{
		libgmsec_perl::LogError($@);
	}

	$self->cleanup();
}

sub reply
{
	my ($self, $request) = @_;
	my $connection = $self->{CONNECTION};
	my $config = $self->{CONFIG};

	# Determine the component which the request is coming from
	my $requestingComponent = $request->getStringField("COMPONENT")->getValue();

	# Set Status Code --  See GMSEC Message Interface Specification (ISD) for available status codes
	my $statusCode = "2"; # 1 = Acknowledgement

	my $replySubject = "GMSEC.MISSION.SAT_ID.RESP."
				. ($requestingComponent ? $requestingComponent : "UNKNOWN-COMPONENT") . "."
				. $statusCode;

	# Create the reply
	my $reply = libgmsec_perl::Message->new($replySubject, $libgmsec_perl::Message::REPLY);

	# Add fields to message
	$reply->addField("ANSWER", "Sure looks like it!");

	$reply->addField("COMPONENT", "gmrpl_cb");

	if ($config->getBooleanValue("ENCRYPT", 0))
	{
		my $field = libgmsec_perl::BooleanField->new("SEC-ENCRYPT", 1);
		$reply->addField($field);
	}

	# Display XML representation of request message
	libgmsec_perl::LogInfo("Prepared reply:\n" . $reply->toXML());

	# Send reply message
	$connection->reply($request, $reply);
}


sub cleanup
{
	my ($self) = @_;

	my $connection = $self->{CONNECTION};

	if ($connection)
	{
		libgmsec_perl::Connection::destroy($connection);
	}

	libgmsec_perl::Connection::shutdownAllMiddlewares();
}

1;

__END__

File: examples/java/gmrpl_cb.java
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/** 
 * @file gmrpl_cb.java
 *
 * An example replier with callback using the Java GMSEC API.
 */


import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.util.*;

import java.util.ArrayList;


class MyReqReplyCallback extends Callback
{
	public void onMessage(Connection conn, Message msg) 
	{
		if (msg.getKind() != Message.MessageKind.REQUEST)
		{
			Log.warning("Received an unexpected message type");
			return;
		}

		try {
			// Display message
			Log.info("onMessage => Received:\n" + msg.toXML());

			// Construct Reply subject.
			String component = msg.getStringField("COMPONENT").getValue();

			// Set Status Code.  
			// See API Interface Specification for available status codes.
			String statusCode = "1";

			StringBuilder subject = new StringBuilder();
			subject.append("GMSEC.MISSION.SAT_ID.RESP.").append(component).append(".").append(statusCode);

			// Create Reply
			Message reply = new Message(subject.toString(), Message.MessageKind.REPLY);

			// Add Fields to replay
			reply.addField("ANSWER",    "Sure looks like it!");
			reply.addField("COMPONENT", "GMRPL");

			// Display reply message
			Log.info("Prepared Reply Message:\n" + reply.toXML());

			// Issue Reply
			conn.reply(msg, reply);

			// Do not destroy the message here
		}
		catch (GMSEC_Exception e) {
			Log.error("onMessage => " + e.toString());
		}
	}
}


public class gmrpl_cb implements Example
{
	static final String DEFAULT_SUBJECT = "GMSEC.>";

	Config     config;
	Connection connection;
	ArrayList<SubscriptionInfo> info = new ArrayList<SubscriptionInfo>();

	gmrpl_cb(Config config) throws ExampleException
	{
		this.config = config;
		Util.initialize(config);

		if (!Util.hasConnectionType(config))
		{
			printUsage();
			throw new ExampleException("Invalid option");
		}
	} 


	private void printUsage()
	{
		Log.error("\nusage: java gmrpl_cb connectionType=<middleware> " +
		          "[ <parameter>=<value> ]\n" +
		          "\n\tNote that the parameter 'connectiontype' is required. "  +
		          "\n\tThe rest of other parameters are optional.\n" +
		          "\n\tserver=<server name> " +
		          "(required if middleware is not bolt/MB locally)" +
		          "\n\tsubject=<subject name>" +
		          "\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
		          "\n\tprog_timeout_s=<timeout period (seconds)>" +
		          "\n\t\tNote: msg_timeout_ms must be specified for prog_timeout_s" +
		          "\n\tloglevel=<log level>" +
		          "\n\tcfgfile=<config_filepath.xml>" +
		          "\n\nFor more information, see API User's Guide\n\n");
	}


	public boolean run() throws GMSEC_Exception
	{
		String subject        = Util.get(config, "SUBJECT", DEFAULT_SUBJECT);
		int    msg_timeout_ms = Util.get(config, "MSG_TIMEOUT_MS", gmsecAPI.WAIT_FOREVER);
		int    prog_timeout_s = Util.get(config, "PROG_TIMEOUT_S", gmsecAPI.WAIT_FOREVER);

		if (msg_timeout_ms == gmsecAPI.WAIT_FOREVER && prog_timeout_s != gmsecAPI.WAIT_FOREVER)
		{
			Log.error("Usage: To use prog_timeout_s, msg_timeout_ms must be defined!");
			return false;
		}

		// Output GMSEC API version
		Log.info(Connection.getAPIVersion());

		connection = Util.openConnection(config);

		// Output middleware version
		Log.info("Middleware version = " + connection.getLibraryVersion());

		// Subscribe
		Log.info("Subscribing to " + subject);
		info.add(connection.subscribe(subject, new MyReqReplyCallback()));

		Log.info("Subscribing to GMSEC.TERMINATE");
		info.add(connection.subscribe("GMSEC.TERMINATE", new MyReqReplyCallback()));


		// Listen
    
		boolean done = false;
		long    prevTime = System.currentTimeMillis();
		long    nextTime = 0;
		int    elapsedTime = 0;	//ms

		while (!done) 
		{
			if (elapsedTime >= (prog_timeout_s * 1000) && prog_timeout_s != gmsecAPI.WAIT_FOREVER)
			{
				done = true;
				Log.info("Program Timeout Reached");
				continue;
			}
    
			Message msg = connection.receive(msg_timeout_ms);

			nextTime     = System.currentTimeMillis();
			elapsedTime += (int) (nextTime - prevTime);
			prevTime     = nextTime;
  
			if (msg == null)
			{
				// This will never print if the parameter is not provided (default 
				// msg_timeout_ms causes GetNextMsg to wait forever)
				Log.info("timeout occurred");
      		}
			else
			{
				done = msg.getSubject().equals("GMSEC.TERMINATE");

				connection.dispatch(msg);
			}
		}

		return true;
	}


	public boolean cleanup() throws GMSEC_Exception
	{
		if (connection != null)
		{
			for(int i = info.size()-1; i >= 0; i-- )
			{
				Log.info("Unsubscribing from " + info.get(i).getSubject());
				connection.unsubscribe(info.get(i));
				info.remove(i);
			}
			Util.closeConnection(connection);
		}

		return true;
	}


	public static void main(String args[]) 
	{
		Example example = null;

		try {
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			example =  new gmrpl_cb(config);
			example.run();
		}
		catch (GMSEC_Exception e) {
			Log.error(e.toString());
		}
		catch (ExampleException e) {
			Log.error(e.toString());
		}
		finally {
			if (example != null )
				Util.cleanup(example);
		}
	}
}
